sktime/tests/test_switch.py

```diff
@@ -100,5 +123,9 @@ def _tests_covering_class_changed(cls):
     # covering the class have changed, then run the test
     cond3 = any(_tests_covering_class_changed(x) for x in cls)
 
-    # run the test if and only if at least one of the conditions 2, 3 are met
-    return cond2 or cond3
+    # Condition 4:
+    # the package requirements for any dependency in pyproject.toml have changed
+    cond4 = any(_is_impacted_by_pyproject_change(x) for x in cls)
+
+    # run the test if and only if at least one of the conditions 2-4 are met
+    return cond2 or cond3 or cond4
```

Comment:
Readability comment, so possibly highly opinionated: if it's a union condition, I'll prefer if you return early as soon as either of cond2, cond3 or cond4 is True instead of checking at end with `or`.

Reply:
is that really better to read?

This way you can easily see what the conditions is precisely, whereas in the alternative case of three returns, you would have to find all returns.

That is, in current case, you need to check a single return line and then the lines where the `condX` are defined, four lines.

In your proposed alterative case, you have to check *every single line in the function*, since you can't be sure you missed a return statement.

Reply:
To me, yes definitely more easily readable. As soon as I read a condition (2 or 3) pass, and I see the corresponding comment saying "if and only if", I want it to end and I get confused why it proceeds to further checks.

I argued with myself whether to post this or not, I sort of expected this response. We can ignore this comment altogether.

Reply:
I would have changed as below. Please ignore long comments and indentation issues, editing in Github UI directly.

existing

```python
    # Condition 1:
    # if any of the required soft dependencies are not present, do not run the test
    if not all(_required_deps_present(x) for x in cls):
        return False
    # otherwise, continue
    # Condition 2:
    # if ONLY_CHANGED_MODULES is on, run the test if and only if
    # any of the modules containing any of the classes in the list have changed
    if ONLY_CHANGED_MODULES:
        cond2 = any(_is_class_changed_or_sktime_parents(x) for x in cls)
    else:
        cond2 = True
    # Condition 3:
    # if the object is an sktime BaseObject, and one of the test classes
    # covering the class have changed, then run the test
    cond3 = any(_tests_covering_class_changed(x) for x in cls)

    # run the test if and only if at least one of the conditions 2, 3 are met
    return cond2 or cond3
    # Condition 4:
    # the package requirements for any dependency in pyproject.toml have changed
    cond4 = any(_is_impacted_by_pyproject_change(x) for x in cls)

    # run the test if and only if at least one of the conditions 2-4 are met
    return cond2 or cond3 or cond4
```

modified

```python
    # Condition 1:
    # if any of the required soft dependencies are not present, do not run the test
    if not all(_required_deps_present(x) for x in cls):
        return False
    # otherwise, continue

    # Condition 2:
    # if ONLY_CHANGED_MODULES is on, run the test if and only if
    # any of the modules containing any of the classes in the list have changed
    if ONLY_CHANGED_MODULES:
        cond2 = any(_is_class_changed_or_sktime_parents(x) for x in cls)
    else:
        cond2 = True

    # at this point we know for sure it'll return True as class itself or at least one of its parent has been changed
    if cond2:
        return True

    # Condition 3:
    # if the object is an sktime BaseObject, and one of the test classes
    # covering the class have changed, then run the test
    cond3 = any(_tests_covering_class_changed(x) for x in cls)

    # at this point we know for sure it'll return True as at least one test is modified
    if cond3:
        return True

    # run the test if and only if at least one of the conditions 2, 3 are met
    return cond2 or cond3
    # Condition 4:
    # the package requirements for any dependency in pyproject.toml have changed
    cond4 = any(_is_impacted_by_pyproject_change(x) for x in cls)

    # at this point tests will run if and only if package dependency of a class has changed but no other code logic change in the definition of itself or its parents or in a test function
    return cond4
```

Reply:
> what did you mean by: " I see the corresponding comment saying "if and only if", I want it to end and I get confused why it proceeds to further checks."

When you wrote the comments for conditions 2-4 and say the test will only run if <some description>, does it not mean that that description being not true must imply test to not run? I read it as such, and since that is not the case here since we are checking more conditions below, that is not clear to me.

Reply:
small question: in your proposed alt, the line ` return cond2 or cond3` seems problematic, since it will always return `False` and prevent `cond4` to be used in a meaningful way.

> When you wrote the comments for conditions 2-4 and say the test will only run if 

I see, this is actually wrong - it comes from a time where only condition 2 existed. This should be corrected. The docstring seems fine though.

Reply:
updated docstring, flow, and comments - should be clearer now?

Reply:
> small question: in your proposed alt, the line  return cond2 or cond3 seems problematic

Yes, sorry, that was a mistake. I'll blame the very small size of Github textbox as my excuse.

> updated docstring, flow, and comments - should be clearer now?

Yes, thank you.